src/mame/drivers/magicard.cpp

```diff
@@ -350,313 +571,656 @@ TODO: check this register,doesn't seem to be 100% correct.
-//  m_screen->screen_vblank().set(m_maincpu, FUNC(scc68070_device::int2_w));
-	m_screen->set_screen_update(FUNC(magicard_state::screen_update_magicard));
+	
+	m_maincpu->i2c_scl_w().set(FUNC(magicard_base_state::cpu_i2c_scl));
+	m_maincpu->i2c_sda_w().set(FUNC(magicard_base_state::cpu_i2c_sda_write));
+	m_maincpu->i2c_sda_r().set(FUNC(magicard_base_state::cpu_i2c_sda_read));
+
+	m_screen->set_raw( CLOCK_A/2, 960, 0, 768, 312, 32, 312);
+	m_screen->set_video_attributes(VIDEO_UPDATE_SCANLINE);
+	SCC66470(config,m_scc66470,CLOCK_A);
+	m_scc66470->set_screen("screen");
+	m_scc66470->irq().set(FUNC(magicard_base_state::scc66470_irq));
+
+	m_screen->set_screen_update(FUNC(magicard_base_state::screen_update_magicard));
+
 
 	PALETTE(config, m_palette).set_entries(0x100);
 	ramdac_device &ramdac(RAMDAC(config, "ramdac", 0, m_palette));
 	ramdac.set_addrmap(0, &magicard_state::ramdac_map);
 
 	SPEAKER(config, "mono").front_center();
+
+	NVRAM(config, "nvram", nvram_device::DEFAULT_ALL_0);
+}
+
+void magicard_state::magicard(machine_config &config)
+{
+	magicard_base(config);
+
+	m_maincpu->set_addrmap(AS_PROGRAM, &magicard_state::magicard_map);
+
+	m_screen->screen_vblank().set(FUNC(magicard_base_state::cpu_int1));
+
 	SAA1099(config, "saa", CLOCK_B).add_route(ALL_OUTPUTS, "mono", 1.0);
+
+	I2C_24C02(config, m_i2cmem);
+	m_i2cmem->write_e0(1);
+
+	DS1207(config, "ds1207");
 }
 
-void magicard_state::hotslots(machine_config &config)
+void magicard_state::magicard_pic54(machine_config &config)
 {
 	magicard(config);
-	m_maincpu->set_addrmap(AS_PROGRAM, &magicard_state::hotslots_map);
 
-	config.device_remove("saa");
+	pic16c54_device &pic(PIC16C54(config, "pic16c54", 3686400)); // correct?
+	pic.read_b().set(FUNC(magicard_base_state::pic_portb_r));
+	pic.write_b().set(FUNC(magicard_base_state::pic_portb_w));
+
+}
+
+void magicard_state::unkte06(machine_config &config)
+{
+	magicard(config);
+
+	pic16c56_device &pic(PIC16C56(config, "pic16c56", 3686400)); // correct?
+	pic.read_b().set(FUNC(magicard_base_state::pic_portb_r));
+	pic.write_b().set(FUNC(magicard_base_state::pic_portb_w));
+}
+
+uint8_t magicard_base_state::pic_portb_r()
+{
+	return (m_sda_state ? 0 : 0x80) | (m_scl_state ? 0 : 0x40);
+}
+
+void magicard_base_state::pic_portb_w(offs_t offset, uint8_t data, uint8_t mask)
+{
+	update_scl(I2C_PIC,((data & mask) | (~mask & 0x40)) & 0x40);
+	update_sda(I2C_PIC,((data & mask) | (~mask & 0x80)) & 0x80);
+}
+
+void hotslots_state::hotslots_base(machine_config &config)
+{
+	magicard_base(config);
+
+	m_maincpu->set_addrmap(AS_PROGRAM, &hotslots_state::hotslots_map);
+
 	YMZ284(config, "ssg", 4000000).add_route(ALL_OUTPUTS, "mono", 1.0);
+	
+	RTC72421(config, m_rtc, XTAL(32'768)).out_int_handler().set(FUNC(hotslots_state::cpu_int1));
+
+	DS2401(config, "serial_id");
+
+	DS1207(config, "ds1207");
+}
+
+void hotslots_state::hotslots(machine_config &config)
+{
+	hotslots_base(config);
+
+	I2C_24C02(config, m_i2cmem);
+	m_i2cmem->write_e0(1);
```

Comment:
You really shouldn’t be setting signal states from machine configuration.  Devices shouldn’t need to anticipate this happening.  It can lead to all kinds of chaos when callbacks aren’t bound, too.

Reply:
OK, didn't realise I shouldn't use that. Actually, looking at other drivers they use set_e0(1) in the same place or as part of the I2C_24C02(....) config which is probably the correct way of doing it ?

Reply:
Can you set the initial signal state at start time instead?